src/hotspot/share/gc/serial/defNewGeneration.cpp

```diff
@@ -601,6 +603,8 @@ void DefNewGeneration::collect(bool   full,
   // Verify that the usage of keep_alive didn't copy any objects.
   assert(heap->no_allocs_since_save_marks(), "save marks have not been newly set.");
 
+  _string_dedup_requests.flush();
```

Comment:
I wonder if placing `_string_dedup_requests.flush();` before weak roots processing (`WeakProcessor::weak_oops_do`) makes more senses. The documentation says `flush` should be called when marking is done; for Serial young GC, it's equivalent to when evacuation is done, which is right after ref-processing.

Reply:
That documentation is mistaken; that's a left-over from an earlier development version that I forgot to update.  Without special additional care, the lifetime also needs to cover finalization.  I have a todo item to file a bug against that comment.